fix: deflake-ify ITSystemTest#queryWatch

BenWhitehead opened the pull request:
queryWatch has been flaky because several things are happening in the
single event stream. This fix splits up the different scenarios being
tested into their own respective tests and updates the assertions to be
more specific to the scenario as well as handling the semantics of the
backend better.

kolea2 commented:
LGTM

BenWhitehead commented:
@schmidt-sebastian Can you get a quick review from you on this PR?

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+  }
+
+  /*
+  Start a listener on a query with an empty result set
```
Please fix (here and everywhere).

BenWhitehead replied:
What fix would you like me to apply here? Turn this into a sentence?

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+
+  /*
+  Starting from a query with empty results creating a new document
+  that matches the query should result in an ADDED event
```
As you fix the comments, do you mind splitting this into two sentences?

BenWhitehead replied:
Will do

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+  that matches the query should result in an ADDED event
+   */
+  @Test
+  public void emptyResults_newDocument_ADDED()
```
This name could use a sprinkle of the Java Style Guide :)

BenWhitehead replied:
This type of scenario is specifically called out in the style guide: http://go/java-style#s5.2.3-method-names
> Underscores may appear in JUnit test method names to separate logical components of the name, with each component written in lowerCamelCase

Is your ask here for me to not use the upper enum names? i.e. `s/ADDED/added/`?

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+    // create a document in our collection that will match the query
+    randomColl.document("doc").set(map("foo", "bar")).get(5, TimeUnit.SECONDS);
+
+    // run our test
```
Nit: I think the "run our test" comments can be removed.

BenWhitehead replied:
Yeah, I'll clean those up.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+
+    ListenerAssertions la = listener.assertions();
+    la.noError();
+    la.eventCountIsAnyOf(Range.closed(1, 2));
```
If you wait for the initial empty event, the number of events is deterministic.

BenWhitehead replied:
Will update

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
+      randomColl.document("doc").set(map("foo", "bar")).get(5, TimeUnit.SECONDS);
```
You don't need to block on this - the write doesn't need to be finished for the listener to fire (the two streams are independent).

BenWhitehead replied:
I see what you're saying, but I'd prefer to keep this here for test reliability sake. There is a difference between "the listener expects an event within 5 seconds from the time the write is complete" to "the listener expects an event within 5 seconds from the time the write is sent". And while the time window there should be small, sometimes events are a bit slow being fired. I've seen a few runs while working on these tests fail in just this small time difference.

schmidt-sebastian replied:
Sounds good.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+
+  /*
+  Starting from a query with empty results, modifying an existing document to
+  match the query should result in an ADDED event
```
Please split into two sentences.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+  match the query should result in an ADDED event
+   */
+  @Test
+  public void emptyResults_modifiedDocument_ADDED()
```
Same comment regarding the style guide.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+    List<ListenerEvent> receivedEvents = listener.receivedEvents;
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
```
This test would be easier to follow if you waited for the initial snapshot. Do you mind updating all tests?

BenWhitehead replied:
I'll make sure all tests wait for the initial event

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+    }
+
+    assertThat(listener.receivedEvents).hasSize(1);
+    ListenerEvent e = listener.receivedEvents.get(0);
```
Nit: Java generally doesn't use abbreviations like this (here and below).

BenWhitehead replied:
I'll rename

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+    }
+  }
+
+  private static final class EventsCDL {
```
Please don't use acronyms. We need to feed the stereotype that Java is much more verbose that C++.

BenWhitehead replied:
Will fix

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,484 @@
+      cdls = new EnumMap<>(DocumentChange.Type.class);
+      cdls.put(DocumentChange.Type.ADDED, new CountDownLatch(addedInitial));
+      cdls.put(DocumentChange.Type.MODIFIED, new CountDownLatch(modifiedInitial));
+      cdls.put(DocumentChange.Type.REMOVED, new CountDownLatch(removedInitial));
```
The way the Watch is used in this test (adding or modifying a document at a time) is actually deterministic. By splitting these updates across types, you add some complexity to the tests that I don't think you need.

BenWhitehead replied:
I tried removing the individual CountDownLatches here and ran into several tests becoming more flaky, so I'd like to leave this for now and spend more time digging into it later.

BenWhitehead commented:
@schmidt-sebastian I've pushed several commits that have addressed most of your comments. (I added some more debugging information to the assertions in the case that they fail to hopefully aid in diagnostics when we see failures. I've seen a few failures working on things, seemingly related to events taking longer than we allow the `await` to block. Is there a defined SLO for events being delivered?)

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+  Attach a listener to a query with empty results.
+  Verify the listener receives an empty event.
+   */
+  @Test
```
Unless this is common in GCloud Java, may I recommend changing this to:


```
  /**
   * Attach a listener to a query with empty results.
   * Verify the listener receives an empty event.
   */
```

If you don't think is is JSDoc worthy then I would move it into the method.

BenWhitehead replied:
Will do

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+      registration.remove();
+    }
+
+    ListenerAssertions la = listener.assertions();
```
Nit: No acronyms.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+      throws InterruptedException, TimeoutException, ExecutionException {
+
+    final Query query = randomColl.whereEqualTo("foo", "bar");
+    QuerySnapshotEventListener listener = new QuerySnapshotEventListener(1, 1, 0, 0);
```
Can you add argument name comments to explain these constants?

BenWhitehead replied:
I've refactored to a builder so we get the names with the arguments.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+
+    EventsCountDownLatch(
+        int nonChangeInitial, int addedInitial, int modifiedInitial, int removedInitial) {
+      cdl = new CountDownLatch(nonChangeInitial);
```
Nit: No acronyms.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+      receivedEvents.add(new ListenerEvent(value, error));
+      if (value != null) {
+        List<DocumentChange> documentChanges = value.getDocumentChanges();
+        for (DocumentChange dc : documentChanges) {
```
Nit: No acronyms.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+    static final class ListenerAssertions {
+      private static final MapJoiner MAP_JOINER = Joiner.on(",").withKeyValueSeparator("=");
+      final List<ListenerEvent> receivedEvents;
+      private final FluentIterable<ListenerEvent> events;
```
Do we need both receivedEvents and events?

BenWhitehead replied:
Nope, leftovers from refactorings. Cleaned up.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,550 @@
+        return builder.toString();
+      }
+
+      private static void debugMessage(StringBuilder builder, QuerySnapshot qs) {
```
Nit: No acronyms.

BenWhitehead commented:
@schmidt-sebastian I've taken care of all the nits, and I update the await conditions to be a multiple of the number of events rather than the firm 5 seconds. This will at least scale our tolerance in the test since.

schmidt-sebastian commented on google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java:
```diff
@@ -0,0 +1,641 @@
+  /**
+   *
+   *
+   * <ol>
```
Nit: Are the two empty lines above from code formatting?

BenWhitehead replied:
They are indeed. I wasn't able to figure out a way to remove them.
